File: wallet/mobile/ios/CoreBindings/MobileNodeSynchronizer.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct MindChainMobileNodeEndpoint: Sendable, Equatable {
    public let baseURL: URL
    public let endpointID: String
    public let controlCluster: String

    public init(baseURL: URL, endpointID: String, controlCluster: String) throws {
        guard baseURL.scheme == "https",
              baseURL.host != nil,
              baseURL.user == nil,
              baseURL.password == nil,
              baseURL.query == nil,
              baseURL.fragment == nil,
              baseURL.path.isEmpty || baseURL.path == "/",
              baseURL.port == nil || baseURL.port == 443,
              Self.hashPattern(endpointID),
              Self.hashPattern(controlCluster) else {
            throw MindChainMobileNodeError.invalidEndpoint
        }
        var components = URLComponents()
        components.scheme = "https"
        components.host = baseURL.host?.lowercased()
        guard let canonical = components.url else {
            throw MindChainMobileNodeError.invalidEndpoint
        }
        self.baseURL = canonical
        self.endpointID = endpointID
        self.controlCluster = controlCluster
    }

    private static func hashPattern(_ value: String) -> Bool {
        value.count == 64 && value.utf8.allSatisfy { byte in
            (48...57).contains(byte) || (97...102).contains(byte)
        }
    }
}

public enum MindChainMobileNodeError: Error, Equatable {
    case invalidEndpoint
    case invalidEndpointSet
    case transportUnavailable
    case invalidResponse
    case responseTooLarge
    case stateUnavailable
}

private final class NoRedirectSessionDelegate: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

/**
 * A bounded mobile node over configured public indexer control clusters. HTTPS
 * transports bytes; the shared Rust core owns chain identity, freshness,
 * control-cluster quorum, ancestry, monotonicity, and durable-state checks.
 */
public actor MindChainMobileNodeSynchronizer {
    private static let maximumResponseBytes = 1_048_576
    private static let maximumEndpoints = 16

    private let endpoints: [MindChainMobileNodeEndpoint]
    private let node: MobileLightNode
    private let stateURL: URL
    private let session: URLSession

    public init(
        chainID: String,
        genesisHash: String,
        maximumFreshnessMilliseconds: UInt64,
        minimumControlClusterQuorum: UInt8,
        endpoints: [MindChainMobileNodeEndpoint],
        stateDirectory: URL
    ) throws {
        guard endpoints.count >= Int(minimumControlClusterQuorum),
              endpoints.count <= Self.maximumEndpoints,
              Set(endpoints.map(\.baseURL)).count == endpoints.count,
              Set(endpoints.map(\.endpointID)).count == endpoints.count else {
            throw MindChainMobileNodeError.invalidEndpointSet
        }
        try FileManager.default.createDirectory(
            at: stateDirectory,
            withIntermediateDirectories: true
        )
        var directoryValues = URLResourceValues()
        directoryValues.isExcludedFromBackup = true
        var protectedDirectory = stateDirectory
        try protectedDirectory.setResourceValues(directoryValues)
        let stateURL = stateDirectory.appendingPathComponent(
            "mindchain-light-node-v1.json",
            isDirectory: false
        )
        let persisted: String?
        if FileManager.default.fileExists(atPath: stateURL.path) {
            let data = try Data(contentsOf: stateURL, options: [.mappedIfSafe])
            guard !data.isEmpty, data.count <= Self.maximumResponseBytes,
                  let value = String(data: data, encoding: .utf8) else {
                throw MindChainMobileNodeError.stateUnavailable
            }
            persisted = value
        } else {
            persisted = nil
        }
        self.endpoints = endpoints
        self.stateURL = stateURL
        self.node = try MobileLightNode(
            chainId: chainID,
            genesisHash: genesisHash,
            apiVersion: "v1",
            maximumFreshnessMs: maximumFreshnessMilliseconds,
            minimumControlClusterQuorum: minimumControlClusterQuorum,
            persistedStateJson: persisted
        )
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 10
        configuration.timeoutIntervalForResource = 20
        configuration.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        configuration.urlCache = nil
        configuration.httpCookieStorage = nil
        configuration.httpShouldSetCookies = false
        configuration.httpMaximumConnectionsPerHost = 2
        self.session = URLSession(
            configuration: configuration,
            delegate: NoRedirectSessionDelegate(),
            delegateQueue: nil
        )
    }

    deinit {
        session.invalidateAndCancel()
    }

    public func snapshot() throws -> MobileNodeSnapshot {
        try node.snapshot()
    }

    public func synchronize() async throws -> MobileNodeSyncOutcome {
        let current = try node.snapshot()
        var observations: [EndpointStatusObservation] = []
        observations.reserveCapacity(endpoints.count)
        for endpoint in endpoints {
            do {
                let status = try await getJSON(
                    endpoint.baseURL.appendingPathComponent("api/status")
                )
                let ancestorHash: String?
                if current.finalizedHeight == 0 {
                    ancestorHash = nil
                } else {
                    let block = try await getJSON(
                        endpoint.baseURL
                            .appendingPathComponent("api/v1/blocks")
                            .appendingPathComponent(String(current.finalizedHeight))
                    )
                    guard let object = try JSONSerialization.jsonObject(with: block) as? [String: Any],
                          let hash = object["hash"] as? String else {
                        throw MindChainMobileNodeError.invalidResponse
                    }
                    ancestorHash = hash
                }
                guard let statusJSON = String(data: status, encoding: .utf8) else {
                    throw MindChainMobileNodeError.invalidResponse
                }
                observations.append(
                    EndpointStatusObservation(
                        endpointId: endpoint.endpointID,
                        controlCluster: endpoint.controlCluster,
                        statusJson: statusJSON,
                        ancestorHash: ancestorHash
                    )
                )
            } catch {
                continue
            }
        }
        let outcome = try node.observeFinalized(observations: observations)
        try persist(outcome.persistedStateJson)
        return outcome
    }

    private func getJSON(_ url: URL) async throws -> Data {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 10
        request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        request.setValue(
            "application/vnd.noos.v1+json, application/json",
            forHTTPHeaderField: "Accept"
        )
        request.setValue("MindChain-Mobile-Node/1", forHTTPHeaderField: "User-Agent")
        let (data, response) = try await session.data(for: request)
        guard data.count <= Self.maximumResponseBytes else {
            throw MindChainMobileNodeError.responseTooLarge
        }
        guard let http = response as? HTTPURLResponse,
              http.statusCode == 200,
              let contentType = http.value(forHTTPHeaderField: "Content-Type")?
                .split(separator: ";", maxSplits: 1).first,
              contentType == "application/json" || contentType == "application/vnd.noos.v1+json" else {
            throw MindChainMobileNodeError.invalidResponse
        }
        return data
    }

    private func persist(_ value: String) throws {
        guard let data = value.data(using: .utf8),
              !data.isEmpty,
              data.count <= Self.maximumResponseBytes else {
            throw MindChainMobileNodeError.stateUnavailable
        }
        try data.write(to: stateURL, options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        var protectedState = stateURL
        try protectedState.setResourceValues(values)
    }
}

File: wallet/mobile/android/core/src/main/java/org/noosphere/wallet/core/MobileNodeSynchronizer.kt
package org.noosphere.wallet.core

import android.content.Context
import android.util.AtomicFile
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL

/** Public indexer in a configured control cluster used by the mobile quorum node. */
data class MobileNodeEndpoint(
    val baseUrl: String,
    val endpointId: String,
    val controlCluster: String,
)

/**
 * Durable state lives in noBackupFilesDir and is replaced through AtomicFile.
 * The Rust core verifies the chain binding, monotonic sequence, checkpoint
 * history, and checksum every time this state is opened.
 */
class MobileNodeStateStore(context: Context, fileName: String = "mindchain-light-node-v1.json") {
    private val file = AtomicFile(File(context.noBackupFilesDir, fileName))

    @Synchronized
    fun read(): String? {
        if (!file.baseFile.isFile) return null
        val bytes = file.openRead().use { input ->
            val output = ByteArrayOutputStream()
            val buffer = ByteArray(8192)
            var total = 0
            while (true) {
                val count = input.read(buffer)
                if (count < 0) break
                total += count
                if (total > MAX_STATE_BYTES) throw IOException("mobile_node_state_too_large")
                output.write(buffer, 0, count)
            }
            output.toByteArray()
        }
        return bytes.toString(Charsets.UTF_8)
    }

    @Synchronized
    fun write(value: String) {
        val bytes = value.toByteArray(Charsets.UTF_8)
        if (bytes.isEmpty() || bytes.size > MAX_STATE_BYTES) {
            throw IOException("mobile_node_state_too_large")
        }
        val output = file.startWrite()
        try {
            output.write(bytes)
            output.fd.sync()
            file.finishWrite(output)
        } catch (error: Exception) {
            file.failWrite(output)
            throw error
        }
    }

    companion object {
        private const val MAX_STATE_BYTES = 1_048_576
    }
}

/**
 * Bounded synchronous synchronization step. Invoke from WorkManager or another
 * background executor, never the main thread. Network transport supplies bytes;
 * all identity, quorum, ancestry, regression, and persistence decisions remain
 * in the shared Rust core.
 */
class MobileNodeSynchronizer(
    context: Context,
    chainId: String,
    genesisHash: String,
    maximumFreshnessMs: ULong,
    minimumControlClusterQuorum: UByte,
    endpoints: List<MobileNodeEndpoint>,
    private val store: MobileNodeStateStore = MobileNodeStateStore(context),
) : AutoCloseable {
    private val endpoints = endpoints.map(::validateEndpoint)
    private val node: MobileLightNode

    init {
        if (this.endpoints.size !in minimumControlClusterQuorum.toInt()..MAX_ENDPOINTS ||
            this.endpoints.map { it.baseUrl }.toSet().size != this.endpoints.size ||
            this.endpoints.map { it.endpointId }.toSet().size != this.endpoints.size
        ) {
            throw IllegalArgumentException("invalid_mobile_node_endpoints")
        }
        node = MobileLightNode(
            chainId = chainId,
            genesisHash = genesisHash,
            apiVersion = "v1",
            maximumFreshnessMs = maximumFreshnessMs,
            minimumControlClusterQuorum = minimumControlClusterQuorum,
            persistedStateJson = store.read(),
        )
    }

    fun snapshot(): MobileNodeSnapshot = node.snapshot()

    @Throws(IOException::class, WalletSdkException::class)
    fun synchronize(): MobileNodeSyncOutcome {
        val current = node.snapshot()
        val observations = endpoints.mapNotNull { endpoint ->
            try {
                val statusJson = getJson(endpoint.baseUrl + "/api/status")
                val ancestorHash = if (current.finalizedHeight == 0UL) {
                    null
                } else {
                    val block = JSONObject(
                        getJson(endpoint.baseUrl + "/api/v1/blocks/${current.finalizedHeight}"),
                    )
                    block.getString("hash")
                }
                EndpointStatusObservation(
                    endpointId = endpoint.endpointId,
                    controlCluster = endpoint.controlCluster,
                    statusJson = statusJson,
                    ancestorHash = ancestorHash,
                )
            } catch (_: IOException) {
                null
            } catch (_: RuntimeException) {
                null
            }
        }
        val outcome = node.observeFinalized(observations)
        store.write(outcome.persistedStateJson)
        return outcome
    }

    override fun close() {
        node.close()
    }

    private fun getJson(url: String): String {
        val connection = URL(url).openConnection() as? HttpURLConnection
            ?: throw IOException("mobile_node_transport_unavailable")
        try {
            connection.instanceFollowRedirects = false
            connection.requestMethod = "GET"
            connection.connectTimeout = CONNECT_TIMEOUT_MS
            connection.readTimeout = READ_TIMEOUT_MS
            connection.setRequestProperty("Accept", "application/vnd.noos.v1+json, application/json")
            connection.setRequestProperty("User-Agent", "MindChain-Mobile-Node/1")
            val status = connection.responseCode
            if (status != HttpURLConnection.HTTP_OK) {
                throw IOException("mobile_node_http_status_$status")
            }
            val contentType = connection.contentType?.substringBefore(';')?.trim()
            if (contentType != "application/json" && contentType != "application/vnd.noos.v1+json") {
                throw IOException("mobile_node_invalid_content_type")
            }
            val body = connection.inputStream.use { input ->
                val output = ByteArrayOutputStream()
                val buffer = ByteArray(8192)
                var total = 0
                while (true) {
                    val count = input.read(buffer)
                    if (count < 0) break
                    total += count
                    if (total > MAX_RESPONSE_BYTES) throw IOException("mobile_node_response_too_large")
                    output.write(buffer, 0, count)
                }
                output.toByteArray()
            }
            return body.toString(Charsets.UTF_8)
        } finally {
            connection.disconnect()
        }
    }

    companion object {
        private const val MAX_ENDPOINTS = 16
        private const val MAX_RESPONSE_BYTES = 1_048_576
        private const val CONNECT_TIMEOUT_MS = 10_000
        private const val READ_TIMEOUT_MS = 10_000

        private fun validateEndpoint(endpoint: MobileNodeEndpoint): MobileNodeEndpoint {
            val uri = try {
                URI(endpoint.baseUrl)
            } catch (error: Exception) {
                throw IllegalArgumentException("invalid_mobile_node_endpoint", error)
            }
            if (uri.scheme != "https" || uri.host.isNullOrEmpty() || uri.userInfo != null ||
                uri.query != null || uri.fragment != null || uri.path !in listOf("", "/") ||
                uri.port !in listOf(-1, 443) || !HASH64.matches(endpoint.endpointId) ||
                !HASH64.matches(endpoint.controlCluster)
            ) {
                throw IllegalArgumentException("invalid_mobile_node_endpoint")
            }
            return endpoint.copy(baseUrl = "https://${uri.host.lowercase()}")
        }

        private val HASH64 = Regex("^[0-9a-f]{64}$")
    }
}
